src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs

```diff
@@ -208,6 +217,19 @@ public async Task ProcessRequestsAsync<TContext>(IHttpApplication<TContext> appl
                     // Call UpdateCompletedStreams() prior to frame processing in order to remove any streams that have exceeded their drain timeouts.
                     UpdateCompletedStreams();
 
+                    if (_keepAlive != null)
+                    {
+                        var state = _keepAlive.ProcessKeepAlive(!result.IsCanceled);
+                        if (state == KeepAliveState.SendPing)
+                        {
+                            await _frameWriter.WritePingAsync(Http2PingFrameFlags.NONE, Http2ConnectionKeepAlive.PingPayload);
```

Comment:
Currently I'm setting the ping sent timestamp before WritePingAsync. Could flow control or something else cause the write to be awaited?